File: layers/bars/imbalance_order_book.jl
using Dates
using PyCall
using DataFrames
using IterTools

import YAML

push!(LOAD_PATH, "C://repos//trade//layers//bars")
import OrderBook

include("C://repos//trade//connector//ts2hdf5//client.jl")
const Client = ClientTsHdf5

datetime = pyimport("datetime")
np = pyimport("numpy")

push!(pyimport("sys")."path", "C://repos//trade")
# build one class that imports all py classes I need
push!(pyimport("sys")."path", "C://repos//trade//common")
Paths = pyimport("paths").Paths
push!(pyimport("sys")."path", "C://repos//trade//common//modules")
# logger = pyimport("logger").logger
push!(pyimport("sys")."path", "C://repos//trade//layers")
BitfinexReader = pyimport("bitfinex_reader").BitfinexReader
push!(pyimport("sys")."path", "C://repos//trade//layers//bars")
# derive_events = pyimport("imbalance_order_book").derive_events


ffill(v) = return v[accumulate(max, [i*!(ismissing(v[i]) | isnan(v[i])) for i in 1:length(v)], init=1)]
isna(v) = return map((i)->(ismissing(i) | isnan(i)), v)

function ix_every_delta(arr, delta)  # no NANs here.
    cumsum_ = vcat([0], cumsum(arr[1:end-1] - arr[2:end]))
    ix_events = Vector()
    actual_deltas = Vector()
    ix_event_old = 1
    while true
        ix_event = 1
        for (ix, el) in enumerate(cumsum_[ix_event_old:end])
            if abs(el) >= delta
                ix_event = ix + ix_event_old
                break
            end
        end
        # ix_event = argmax(abs.(cumsum_) .>= delta)
        if ix_event == 1
            break
        else
            actual_delta = cumsum_[ix_event]
            push!(ix_events, ix_event)
            push!(actual_deltas, actual_delta)
            cumsum_ .-= actual_delta
            # cumsum_[1:ix_event] .= 0
            ix_event_old = ix_event
        end
    end
    return ix_events, actual_deltas
end

function invert_lt_zero_ratio!(v)
    ix_ask_greater = v .< 1  # why 1 not 0
    v[ix_ask_greater] .= 1 ./ v[ix_ask_greater]
    return v
end

function derive_events(dfo; level=30)
    df = filter!(r -> r.level <= level, dfo)
    df = combine(DataFrames.groupby(df, ["timestamp", "side", "level"]), :size=>last, :count=>last)
    ix_ts = unique(df.timestamp) |> sort
    map_side = Dict([(-1, 1), (1, 2)])

    book = zeros(Union{Missing, Float64}, length(ix_ts), 2, level, 2)
    book .= missing
    for side in [-1, 1]
        for level in 1:30
            if length(df[(df.side.==side) .& (df.level.==level), "size_last"]) > 0
                book[:, map_side[side], level, 1] = df[(df.side.==side) .& (df.level.==level), "size_last"]
                book[:, map_side[side], level, 2] = df[(df.side.==side) .& (df.level.==level), "count_last"]
            end
        end
    end
    return book
end


function main()
    @time begin
    settings = YAML.load_file(Paths.layer_settings)
    for exchange in keys(settings)
        println("$(exchange)")
        for (asset, params) in pairs(settings[exchange])
            if asset != "ethusd"
                continue
            end
            println("Loading order book for $(exchange) - $(asset)")
            params = get(params, "order book", Dict())
            delta_size_ratio = get(params, "delta_size_ratio", 0)
            start = Date(2022, 2, 7)
            end_ = Date(2022, 2, 7)
            for i = 1:1 + Dates.value(end_-start)
                dt = start + Day(i)
                println("Running $(dt)")
                df_py::PyObject = BitfinexReader.load_quotes(asset, datetime.datetime(2022, 2, 8), datetime.datetime(2022, 2, 8))
                df::DataFrame = DataFrame(
                    timestamp=Nanosecond.(df_py.get("timestamp").astype(np.int64)) + DateTime(1970),
                    price=df_py.get("price").values,
                    size=df_py.get("size").values,
                    count=df_py.get("count").values,
                    side=df_py.get("side").values,
                )
                df.timestamp .= df.timestamp[1]
                ob = OrderBook.Book(df, get(params, "level_from_price_pct", 0), 0.1)  # need to change or infer this 1.
                level = 30
                println("$(asset) - Summarizing $(level) order book levels")
                df = OrderBook.create_order_book!(ob)
                arr = derive_events(df)

                ix_valid = 1
                for (side, level) in product(1:2, 1:level)
                    println("Side: $(side) - Level: $(level)")
                    
                    arr[:, side, level, 1] = ffill(arr[:, side, level, 1])
                    arr[:, side, level, 2] = ffill(arr[:, side, level, 2])

                    ix_valid = max(argmin(isna(arr[:, side, level, 1])), ix_valid)
                    ix_valid = max(argmin(isna(arr[:, side, level, 2])), ix_valid)
                end

                # arr, v_ts = arr[ix_valid:end, :, :, :], (Nanosecond.(map(((i)->i.astype(np.int64)), df.timestamp)) + DateTime(1970))[ix_valid:end]
                arr, v_ts = arr[ix_valid:end, :, :, :], df.timestamp[ix_valid:end]
                arr[isnan.(arr) .| ismissing.(arr)] .= 0

                alpha = 2/(level + 1)
                weights = reshape(map((i)->(1-alpha)^i, 0:level-1), (1, 1, level, 1))

                arr = reshape(sum(arr .* weights, dims=3), (:, 2, 2))
                m_size, m_count = arr[:, :, 1], arr[:, :, 2]
                min_pos_cnt = minimum(m_count[m_count .> 0])
                m_count = ifelse.(m_count .== 0, min_pos_cnt, m_count)  # avoid divide by 0 error
                m_count = ifelse.(m_count .== 0, min_pos_cnt, m_count)  # avoid divide by 0 errorda
                min_pos_size = minimum(m_size[m_size .> 0])
                m_size = ifelse.(m_size .== 0, min_pos_size, m_size)  # avoid divide by 0 error

                count_ratio = m_count[:, 1] ./ abs.(m_count[:, 2])
                size_ratio = m_size[:, 1] ./ abs.(m_size[:, 2])  # bid|buy / ask|sell
                v_count_ratio = invert_lt_zero_ratio!(count_ratio)
                v_size_ratio = invert_lt_zero_ratio!(size_ratio)
                v_size_net = m_size[:, 1] + m_size[:, 2]
                v_count_net = m_count[:, 1] - m_count[:, 2]  # bid - ask
                
                println("Get Delta steps")
                @time ix_events, actual_deltas = ix_every_delta(v_size_ratio, delta_size_ratio)
                v_ts = v_ts[ix_events]
                @time ix_ts_unique_last = findlast.(isequal.(unique(v_ts)), [v_ts])
                println("Ingesting Size, Count Ratios and Net from total $(length(v_size_ratio)), 
                        reduced to # events $(length(v_ts)) -> # unique last ts: $(length(ix_ts_unique_last))")
                v_ts = v_ts[ix_ts_unique_last]
                # somehow need to ensure here only 1 value per timestamp
                for (information, v) in pairs(Dict(
                    "bid_buy_size_imbalance_ratio" => v_size_ratio,
                    "bid_buy_count_imbalance_ratio" => v_count_ratio,
                    "bid_buy_size_imbalance_net" => v_size_net,
                    "bid_buy_count_imbalance_net" => v_count_net,
                ))
                    Client.upsert(
                        Dict(
                            "measurement_name" => "order book",
                            "exchange" => exchange,
                            "asset" => asset,
                            "information" => information,
                            "unit" => "size_ewm_sum",
                            "delta_size_ratio" => delta_size_ratio
                        ),
                        hcat(v_ts, v[ix_events][ix_ts_unique_last])
                    )
                end
            end
        end
    end
    println("Done")
end
end

main()

File: layers/bars/imbalance_order_book.py
from typing import List

import yaml
import numpy as np
import datetime
import pandas as pd

from functools import lru_cache
from itertools import product

from numba import jit

from common.paths import Paths
from connector.influxdb.influxdb_wrapper import influx
from common.modules.logger import logger


def apply_best_bid_ask(price: np.ndarray, side: np.ndarray, size: np.ndarray):
    active_bids = set()
    active_asks = set()
    n = len(price)
    bbid = price[np.argmax(side == 1)]
    bask = price[np.argmax(side == -1)]
    best_bids = np.empty(n)
    best_asks = np.empty(n)
    for i in range(n):
        p = price[i]
        c = side[i]
        if c == 0:
            if size[i] > 0:
                if p in active_bids:
                    active_bids.remove(p)
                if p > bbid:
                    if active_bids:
                        bbid = max(active_bids)
                    else:
                        bbid = 0

            elif size[i] < 0:
                if p in active_asks:
                    active_asks.remove(p)

                if p < bask:
                    if active_asks:
                        bask = min(active_asks)
                    else:
                        bask = 99999
        else:
            if size[i] > 0:
                active_bids.add(p)
                if p >= bbid:
                    bbid = p
            elif size[i] < 0:
                active_asks.add(p)
                if p <= bask:
                    bask = p

        if bbid >= bask:  # something went wrong. make it consistent
            if size[i] > 0:  # is bid
                for n in [n for n in active_asks if n <= bbid]:
                    active_asks.remove(n)
                bask = min(active_asks) if active_asks else 99999
            if size[i] < 0:  # is ask
                for n in [n for n in active_bids if n >= bask]:
                    active_bids.remove(n)
                bbid = max(active_bids) if active_bids else 0

        best_bids[i] = bbid
        best_asks[i] = bask
        if i % 1000000 == 0:
            print(i)
    return best_bids, best_asks


class OrderBook:
    """
    - Multi dimensional frame with dimensions: time, side, level    ( potentially another to get size of each bid/ask within level)!
        created from a stream of ticks
    - Best bid ask frame.
    """
    def __init__(self, df_quotes, level_from_price_pct):
        self.df_quotes = df_quotes
        self.level_from_price_pct = level_from_price_pct

    @property
    @lru_cache()
    def level_distance(self):
        res = (self.df_quotes['price'].shift(1) - self.df_quotes['price']).abs()
        return res[res != 0].min()

    def create_order_book(self):
        df = self.df_quotes
        assert sorted(df['side'].unique()) == [-1, 1], 'Side is not fully determined. Infer from BBAB and price'
        # Count may only be availbe with Bitfinex at the moment. Enrich if necessary. Emptied levels, count 0, mean to be ignore for best bid determination, but
        # need information to accurately encode when levels where emptied and filled
        a, b = apply_best_bid_ask(df['price'].values, np.where(df['count'] == 0, 0, df['side']), df['size'].values)
        df['best_bid'] = a
        df['best_ask'] = b

        df['best_bid'] = df['best_bid'].ffill()
        df.loc[df.index[df['best_bid'] == 0], 'best_bid'] = None

        df['best_ask'] = df['best_ask'].ffill()
        df.loc[df.index[df['best_ask'] == 99999], 'best_ask'] = None
        df['best_ask'] = df['best_ask'].bfill()
        df['best_bid'] = df['best_bid'].bfill()
        # df = df.iloc[1000:].reset_index(drop=True)
        df = df[~df['price'].isna()]
        df = df[~df['size'].isna()]
        df = self.impute_nan_count(df)
        assert df.isna().sum().sum() == 0, 'NANs at this step. why?'
        assert (df['best_ask'] < df['best_bid']).sum() == 0
        logger.info(f'Null price rows: {df["price"].isna().sum()}')
        # Add Level

        ix_drop_no_level_land = df.index[np.where((df['best_bid'] < df['price']) & (df['price'] < df['best_ask']))]
        if len(ix_drop_no_level_land) > 0:
            logger.info(f'''Dropping {len(ix_drop_no_level_land)} order book levels that are in between best ask and best bid.''')
            df = df.drop(ix_drop_no_level_land).reset_index(drop=True)

        # Assign side to emptied levels
        ix_level_empty = df.index[df['count'] == 0]
        df.loc[ix_level_empty, 'size'] = 0
        assert df['size'][df['count'] == 0].sum() == 0, 'Size should be 0 whenever emtpy order book level / count is 0.'

        # Might be better to drop these rather than reassigning. Could be due to data coming in wrong sequence.
        ix_wrong_side_ask = df.index[(df['side'] == -1) & (df['price'] < df['best_ask'])]
        ix_wrong_side_bid = df.index[(df['side'] == 1) & (df['price'] > df['best_bid'])]
        if len(ix_wrong_side_bid) + len(ix_wrong_side_ask) > 0:
            logger.info(f'''Reassigning Inconsistent / Wrong side level BID: {len(ix_wrong_side_bid)} - ASK: {len(ix_wrong_side_ask)}. Investigate if inconsistency count is high ''')
            df.loc[ix_wrong_side_bid, 'side'] *= -1
            df.loc[ix_wrong_side_ask, 'side'] *= -1

        # Filled Levels
        ix_ask = df.index[df['side'] == -1]
        ix_bid = df.index[df['side'] == 1]
        df['level'] = None
        df.loc[ix_ask, 'level'] = ((df.loc[ix_ask, 'price'] - df.loc[ix_ask, 'best_ask']) / self.level_distance + 1).astype(int)
        df.loc[ix_bid, 'level'] = ((df.loc[ix_bid, 'price'] - df.loc[ix_bid, 'best_bid']) / self.level_distance - 1).astype(int)

        assert df.loc[ix_ask, 'level'].min() >= 1
        assert df.loc[ix_bid, 'level'].max() <= 0
        assert df['level'].isna().sum() == 0

        df['level'] = df['level'].astype(int).abs()
        self.df_quotes = df = df.set_index(['timestamp', 'side', 'level'])

    @staticmethod
    def impute_nan_count(df: pd.DataFrame) -> pd.DataFrame:
        if df.isna().sum()['count'] > 0:
            ix_zero = df.index[(df['count'].isna()) & (df['size'].abs() == 1)]
            df.loc[ix_zero, 'count'] = 0
            ix_non_zero = df.index[(df['count'].isna()) & (df['size'].abs() != 1)]
            df.loc[ix_non_zero, 'count'] = 1
            logger.info(f'Imputed {len(ix_zero) + len(ix_non_zero)} count values.')
            return df
        else:
            return df

    @property
    @lru_cache()
    def level(self) -> int:
        return min(int((self.df_quotes['price'] * self.level_from_price_pct / 100).max() / self.level_distance), 300)

    def derive_events(self):
        df = self.df_quotes
        df = df.loc[(slice(None), slice(None), list(range(self.level))), :]
        df = df.groupby(['timestamp', 'side', 'level']).last()
        # cumulative race through all rows. final state must include every single level
        ix_ts = df.index.get_level_values('timestamp').unique()
        book = pd.DataFrame(None, index=pd.MultiIndex.from_product([
                                ix_ts,
                                [1, -1],  # side
                                range(self.level)
                            ], names=["timestamp", "side", 'level']),
                            columns=['size', 'count'], dtype=float)
        # book2 = np.empty(shape=(len(df.index.get_level_values('timestamp').unique()), 2, self.level, 2))
        logger.info('Inserting values into empty book frame')
        book.loc[df.index] = df[['size', 'count']]
        df = None
        arr = book.values.reshape(len(ix_ts), 2, self.level, 2)
        book = None
        ix_valid = 0
        logger.info('FFill each level')
        for side, level in product([0, 1], range(self.level)):
            # print(f'Side: {side} - Level: {level}')
            arr[:, side, level, 0] = ffill(arr[:, side, level, 0])
            ix_valid = max(np.argmin(np.isnan(arr[:, side, level, 0])), ix_valid)
            arr[:, side, level, 1] = ffill(arr[:, side, level, 1])
            ix_valid = max(np.argmin(np.isnan(arr[:, side, level, 1])), ix_valid)
        arr, ts = arr[ix_valid:], ix_ts[ix_valid:]
        arr = np.nan_to_num(arr)
        return arr, ts


def ffill(arr: np.ndarray) -> np.ndarray:
    # same technique - ffill across all levels and sides
    return arr[np.maximum.accumulate(np.where(~np.isnan(arr), np.arange(arr.shape[0]), 0))]


# @jit(nopython=False)
def ix_every_delta(arr: np.array, delta: float) -> (List, List):
    cumsum = np.array([0] + (arr[:-1] - arr[1:]).cumsum().tolist())
    ix_events = []
    actual_deltas = []
    while True:
        ix_event = np.argmax(np.abs(cumsum) >= delta)
        if ix_event == 0:
            break
        else:
            actual_delta = cumsum[ix_event]
            ix_events.append(ix_event)
            actual_deltas.append(actual_delta)
            cumsum -= actual_delta
            cumsum[:ix_event] = 0
    return ix_events, actual_deltas


def invert_lt_zero_ratio(ps: pd.Series) -> pd.Series:
    ix_ask_greater = ps.index[ps < 1]  # i
    ps.loc[ix_ask_greater] = 1 / ps.loc[ix_ask_greater]
    return ps


if __name__ == '__main__':
    with open(Paths.layer_settings, "r") as stream:
        settings = yaml.safe_load(stream)
    for exchange in settings.keys():
        for asset, params in settings[exchange].items():
            if asset not in ['solusd']:  # continue from march 8
            # if asset not in ['xrpusd']:
                continue

            logger.info(f'Loading order book for {exchange} - {asset}')
            params = params.get('order book', {})
            delta_size_ratio = params.get('delta_size_ratio')
            from layers.bitfinex_reader import BitfinexReader
            start = datetime.datetime(2022, 2, 7)
            end = datetime.datetime(2022, 3, 13)
            for i in range(1 + (end-start).days):
                dt = start + datetime.timedelta(days=i)
                logger.info(f'Running {dt}')
                df = BitfinexReader.load_quotes(sym=asset, start=dt, end=dt)
                ob = OrderBook(df, level_from_price_pct=params.get('level_from_price_pct'))
                level = ob.level
                logger.info(f'{asset} - Summarizing {level} order book levels')
                ob.create_order_book()
                arr, ix_ts = ob.derive_events()

                alpha = 2/(level + 1)
                weights = np.array([(1-alpha)**i for i in range(level)]).reshape(1, 1, level, 1)

                arr = (arr * weights).sum(axis=2)
                size, count = arr[:, :, 0], arr[:, :, 1]

                count_ratio = (count[:, 0] / np.abs(count[:, 1]))
                size_ratio = (size[:, 0] / np.abs(size[:, 1]))  # bid|buy / ask|sell
                ps_count_ratio = invert_lt_zero_ratio(pd.Series(count_ratio, index=ix_ts, name='count_ratio'))
                ps_size_ratio = invert_lt_zero_ratio(pd.Series(size_ratio, index=ix_ts, name='size_ratio'))
                ps_size_net = pd.Series(size[:, 0] + size[:, 1], index=ix_ts, name='size_net')
                ps_count_net = pd.Series(count[:, 0] - count[:, 1], index=ix_ts, name='count_net')  # bid - ask

                ix_events, actual_deltas = ix_every_delta(ps_size_ratio.values, delta=delta_size_ratio)
                ts_events = ix_ts[ix_events].unique()
                logger.info(f'Ingesting Size, Count Ratios and Net for {len(ix_events)} unique: {len(ts_events)}')

                for information, ps in {
                    'bid_buy_size_imbalance_ratio': ps_size_ratio,
                    'bid_buy_count_imbalance_ratio': ps_count_ratio,
                    'bid_buy_size_imbalance_net': ps_size_net,
                    'bid_buy_count_imbalance_net': ps_count_net,
                }.items():
                    influx.write(
                        record=ps.loc[ts_events].groupby(level=0).last(),
                        data_frame_measurement_name='order book',
                        data_frame_tag_columns={**{
                            'exchange': exchange,
                            'asset': asset,
                            'information': information,
                            'unit': 'size_ewm_sum',
                            'levels': level,
                            'alpha': alpha,
                            'delta_size_ratio': delta_size_ratio
                        }},
                    )
    logger.info('Done')

